### ISSUE TITLE
provided Map and Sets will be recreated in Subcomponents

### ISSUE THREAD

**Issue by kohlhaas** [google/dagger#257]
This component does create extra providers for set and maps in the subcomponent, although they are not used.

If you have many or large provided maps or sets and your subcomponent is used with a scope that is created on request, dagger recreates all the maps and sets on each request, although they are never injected.

``` java
@javax.inject.Singleton @dagger.Component(modules = MapBugSubcomponent.MyModule.class) public interface MapBugSubcomponent
{
    MainInjectable getMainInjectable();

    MySubComponent getSubComponent();

    @dagger.Subcomponent interface MySubComponent
    {
        // SubComponent does not need any provided Map and Set but dagger generates unused map and set providers for the Subcomponent
        // this only happens with map and sets. The integer provider from the main component is correctly reused without regeneration
        // in the subcomponent
        SubComponentInjectable getSubInjectable();
    }

    // Provides all the objects for MainInjectable
    @dagger.Module class MyModule
    {
        @javax.inject.Singleton @dagger.Provides Integer provideInt()
        {
            return 1;
        }

        @javax.inject.Singleton @dagger.Provides(type = dagger.Provides.Type.SET_VALUES) java.util.Set<String> provideSet()
        {
            return java.util.Collections.singleton("SetEntry");
        }

        @MyMapKey("1") @javax.inject.Singleton @dagger.Provides(type = dagger.Provides.Type.MAP) String provideEntry()
        {
            return "MapEntry";
        }
    }

    @dagger.MapKey(unwrapValue = true) public @interface MyMapKey
    {
        String value();
    }

    class MainInjectable
    {
        @javax.inject.Inject MainInjectable(java.util.Map<String, String> map, java.util.Set<String> set, Integer integer)
        {
        }
    }

    class SubComponentInjectable
    {
        @javax.inject.Inject SubComponentInjectable(Integer integer)
        {
        }
    }
}
```

Note that in the generated code

```
private Provider<Map<String, Provider<String>>> mapOfStringAndProviderOfStringProvider;
private Provider<Map<String, String>> mapOfStringAndStringProvider;
private Provider<Set<String>> setOfStringProvider;
```

are recreated although they are not used in the Objects created by the subcomponent.
Also note that provideIntProvider that is used in the subcomponent is not recreated and correctly used from the main component. 

``` java
@Generated("dagger.internal.codegen.ComponentProcessor")
public final class DaggerMapBugSubcomponent implements MapBugSubcomponent {
  private Provider<String> mapOfStringAndProviderOfStringContribution1;
  private Provider<Map<String, Provider<String>>> mapOfStringAndProviderOfStringProvider;
  private Provider<Map<String, String>> mapOfStringAndStringProvider;
  private Provider<Set<String>> setOfStringContribution1Provider;
  private Provider<Set<String>> setOfStringProvider;
  private Provider<Integer> provideIntProvider;
  private Provider<MainInjectable> mainInjectableProvider;

  private DaggerMapBugSubcomponent(Builder builder) {  
    assert builder != null;
    initialize(builder);
  }

  public static Builder builder() {  
    return new Builder();
  }

  public static MapBugSubcomponent create() {  
    return builder().build();
  }

  private void initialize(final Builder builder) {  
    this.mapOfStringAndProviderOfStringContribution1 = ScopedProvider.create(MapBugSubcomponent$MyModule_ProvideEntryFactory.create(builder.myModule));
    this.mapOfStringAndProviderOfStringProvider = MapProviderFactory.<String, String>builder(1)
        .put("1", mapOfStringAndProviderOfStringContribution1)
        .build();
    this.mapOfStringAndStringProvider = MapFactory.create(mapOfStringAndProviderOfStringProvider);
    this.setOfStringContribution1Provider = ScopedProvider.create(MapBugSubcomponent$MyModule_ProvideSetFactory.create(builder.myModule));
    this.setOfStringProvider = SetFactory.create(setOfStringContribution1Provider);
    this.provideIntProvider = ScopedProvider.create(MapBugSubcomponent$MyModule_ProvideIntFactory.create(builder.myModule));
    this.mainInjectableProvider = MapBugSubcomponent$MainInjectable_Factory.create(mapOfStringAndStringProvider, setOfStringProvider, provideIntProvider);
  }

  @Override
  public MainInjectable getMainInjectable() {  
    return mainInjectableProvider.get();
  }

  @Override
  public MySubComponent getSubComponent() {  
    return new MySubComponentImpl();
  }

  public static final class Builder {
    private MyModule myModule;

    private Builder() {  
    }

    public MapBugSubcomponent build() {  
      if (myModule == null) {
        this.myModule = new MyModule();
      }
      return new DaggerMapBugSubcomponent(this);
    }

    public Builder myModule(MyModule myModule) {  
      if (myModule == null) {
        throw new NullPointerException("myModule");
      }
      this.myModule = myModule;
      return this;
    }
  }

  private final class MySubComponentImpl implements MySubComponent {
    private Provider<SubComponentInjectable> subComponentInjectableProvider;
    private Provider<Map<String, Provider<String>>> mapOfStringAndProviderOfStringProvider;
    private Provider<Map<String, String>> mapOfStringAndStringProvider;
    private Provider<Set<String>> setOfStringProvider;

    private MySubComponentImpl() {  
      initialize();
    }

    private void initialize() {  
      this.subComponentInjectableProvider = MapBugSubcomponent$SubComponentInjectable_Factory.create(DaggerMapBugSubcomponent.this.provideIntProvider);
      this.mapOfStringAndProviderOfStringProvider = MapProviderFactory.<String, String>builder(1)
          .put("1", mapOfStringAndProviderOfStringContribution1)
          .build();
      this.mapOfStringAndStringProvider = MapFactory.create(mapOfStringAndProviderOfStringProvider);
      this.setOfStringProvider = SetFactory.create(setOfStringContribution1Provider);
    }

    @Override
    public SubComponentInjectable getSubInjectable() {  
      return subComponentInjectableProvider.get();
    }
  }
}
```

**Comment by ronshapiro** [comment]
@netdpb is this still an issue?

**Comment by kohlhaas** [comment]
Seem's to be fixed in dagger 2.7.  See new code

``` java
@Generated(
  value = "dagger.internal.codegen.ComponentProcessor",
  comments = "https://google.github.io/dagger"
)
public final class DaggerMapBugSubcomponent implements MapBugSubcomponent {
  private Provider<String> provideEntryProvider;

  private Provider<Map<String, Provider<String>>> mapOfStringAndProviderOfStringProvider;

  private Provider<Map<String, String>> mapOfStringAndStringProvider;

  private Provider<Set<String>> provideSetProvider;

  private Provider<Set<String>> setOfStringProvider;

  private Provider<Integer> provideIntProvider;

  private Provider<MapBugSubcomponent.MainInjectable> mainInjectableProvider;

  private DaggerMapBugSubcomponent(Builder builder) {
    assert builder != null;
    initialize(builder);
  }

  public static Builder builder() {
    return new Builder();
  }

  public static MapBugSubcomponent create() {
    return builder().build();
  }

  @SuppressWarnings("unchecked")
  private void initialize(final Builder builder) {

    this.provideEntryProvider =
        DoubleCheck.provider(
            MapBugSubcomponent_MyModule_ProvideEntryFactory.create(builder.myModule));

    this.mapOfStringAndProviderOfStringProvider =
        MapProviderFactory.<String, String>builder(1).put("1", provideEntryProvider).build();

    this.mapOfStringAndStringProvider = MapFactory.create(mapOfStringAndProviderOfStringProvider);

    this.provideSetProvider =
        DoubleCheck.provider(
            MapBugSubcomponent_MyModule_ProvideSetFactory.create(builder.myModule));

    this.setOfStringProvider =
        SetFactory.<String>builder(0, 1).addCollectionProvider(provideSetProvider).build();

    this.provideIntProvider =
        DoubleCheck.provider(
            MapBugSubcomponent_MyModule_ProvideIntFactory.create(builder.myModule));

    this.mainInjectableProvider =
        MapBugSubcomponent_MainInjectable_Factory.create(
            mapOfStringAndStringProvider, setOfStringProvider, provideIntProvider);
  }

  @Override
  public MapBugSubcomponent.MainInjectable getMainInjectable() {
    return mainInjectableProvider.get();
  }

  @Override
  public MapBugSubcomponent.MySubComponent getSubComponent() {
    return new MySubComponentImpl();
  }

  public static final class Builder {
    private MapBugSubcomponent.MyModule myModule;

    private Builder() {}

    public MapBugSubcomponent build() {
      if (myModule == null) {
        this.myModule = new MapBugSubcomponent.MyModule();
      }
      return new DaggerMapBugSubcomponent(this);
    }

    public Builder myModule(MapBugSubcomponent.MyModule myModule) {
      this.myModule = Preconditions.checkNotNull(myModule);
      return this;
    }
  }

  private final class MySubComponentImpl implements MapBugSubcomponent.MySubComponent {
    private Provider<MapBugSubcomponent.SubComponentInjectable> subComponentInjectableProvider;

    private MySubComponentImpl() {
      initialize();
    }

    @SuppressWarnings("unchecked")
    private void initialize() {

      this.subComponentInjectableProvider =
          MapBugSubcomponent_SubComponentInjectable_Factory.create(
              DaggerMapBugSubcomponent.this.provideIntProvider);
    }

    @Override
    public MapBugSubcomponent.SubComponentInjectable getSubInjectable() {
      return subComponentInjectableProvider.get();
    }
  }
}
```
